src/datatypes/numeric.js

```diff
@@ -1,58 +1,82 @@
 const { PartialReadError } = require('../utils')
 
+class BigIntExtended extends Array {
+  valueOf () { return BigInt.asIntN(64, BigInt(this[0]) << 32n) | BigInt.asUintN(32, BigInt(this[1])) }
+}
+
 function readI64 (buffer, offset) {
   if (offset + 8 > buffer.length) { throw new PartialReadError() }
+  const val = buffer.readBigInt64BE(offset)
   return {
-    value: [buffer.readInt32BE(offset), buffer.readInt32BE(offset + 4)],
+    value: new BigIntExtended(buffer.readInt32BE(offset), buffer.readInt32BE(offset + 4)),
     size: 8
   }
 }
 
 function writeI64 (value, buffer, offset) {
-  buffer.writeInt32BE(value[0], offset)
-  buffer.writeInt32BE(value[1], offset + 4)
+  if (typeof value.valueOf() === 'bigint') {
```

Comment:
This way to check if it's a bigint seems costly.
It's doing the conversion to bigint and not just checking the type
Can't this just check instanceof BigIntExtended or BigInt instead ?

Reply:
ah yeah, the shifting could be an issue, updated to instanceof